config-model-generator/src/main/java/io/strimzi/build/kafka/metadata/KafkaConfigModelGenerator.java

```diff
@@ -84,17 +84,36 @@ private static Map<String, ConfigModel> configs() throws NoSuchMethodException,
                 descriptor.setValues(enumer(key.validator));
             } else if (key.validator instanceof ConfigDef.ValidList) {
                 descriptor.setItems(validList(key));
-            } else if (key.validator instanceof ApiVersionValidator$) {
-                Iterator<ApiVersion> iterator = ApiVersion$.MODULE$.allVersions().iterator();
+            } else if (key.validator instanceof MetadataVersionValidator) {
+                // Versions for Kafka 3.3.0 and newer are in MetadataVersion
+                // We extract it from there
+                Iterator<MetadataVersion> iterator = Arrays.stream(MetadataVersion.VERSIONS).iterator();
                 LinkedHashSet<String> versions = new LinkedHashSet<>();
                 while (iterator.hasNext()) {
-                    ApiVersion next = iterator.next();
-                    ApiVersion$.MODULE$.apply(next.shortVersion());
+                    MetadataVersion next = iterator.next();
                     versions.add(Pattern.quote(next.shortVersion()) + "(\\.[0-9]+)*");
-                    ApiVersion$.MODULE$.apply(next.version());
                     versions.add(Pattern.quote(next.version()));
                 }
                 descriptor.setPattern(String.join("|", versions));
+            } else if (key.validator != null && "class kafka.api.ApiVersionValidator$".equals(key.validator.getClass().toString())) {
+                // Versions for Kafka 3.2.x and older are in ApiVersions. We cannot use this class because it does not
+                // exist anymore in Kafka 3.3.0. So we extract the versions from Kafka 3.3.0, but we filter them to have
+                // the short version equal or less the version for which we generate the config. This should filter out
+                // the older versions and keep only the valid versions for given release.
+                Iterator<MetadataVersion> iterator = Arrays.stream(MetadataVersion.VERSIONS).iterator();
+                LinkedHashSet<String> versions = new LinkedHashSet<>();
+                while (iterator.hasNext()) {
+                    MetadataVersion next = iterator.next();
+                    if (compareDottedVersions(version, next.shortVersion()) >= 0) {
+                        versions.add(Pattern.quote(next.shortVersion()) + "(\\.[0-9]+)*");
+                        versions.add(Pattern.quote(next.version()));
+                    }
+                }
+                descriptor.setPattern(String.join("|", versions));
+
+                //continue;
```

Comment:
@tombentley This is a bit of a hack which works around the change https://github.com/apache/kafka/commit/8245c9a3d5af2ad891844194a5fa281af471b568 => it refactored the `ApiVersion` class into `MetadataVersion`. That makes it hard to handle the configuration for the different Kafka versions with some using the old and some using the new. Do you have any better idea how to handle it then this? This seems to work, but it is a bit hack.